staging/src/k8s.io/apiserver/pkg/apis/audit/types.go

```diff
@@ -0,0 +1,169 @@
+	// Source IP, from where the request originates.
+	// +optional
+	SourceIP string
+	// Object reference this request is targeted at.
+	// Does not apply for List-type requests, or non-resource requests.
+	// +optional
+	ObjectRef *api.ObjectReference
+	// The response status, populated even when the ResponseObject is not a Status type.
+	// For successful responses, this will only include the Code and StatusSuccess.
+	// For non-status type error responses, this will be auto-populated with the error Message.
+	// +optional
+	ResponseStatus *metav1.Status
+
+	// API object from the request. The RequestObject is recorded as-is in the request, prior to
+	// version conversion, defaulting, admission or merging. It is an external versioned object type,
+	// and may not be a valid object on its own.  Omitted for non-resource requests.  Only logged at
+	// RequestObject Level and higher.
+	// +optional
+	RequestObject runtime.Object
+	// API object returned in the response. The ResponseObject is recorded after conversion to the
+	// external type.  Omitted for non-resource requests.  Only logged at ResponseObject Level and
+	// higher.
+	// +optional
+	ResponseObject runtime.Object
+}
+
+// EventList is a list of audit Events.
+type EventList struct {
+	metav1.TypeMeta
+	// +optional
+	metav1.ListMeta
+
+	Events []Event
+}
+
+// Policy defines the configuration of audit logging, and the rules for how different request
+// categories are logged.
+type Policy struct {
+	metav1.TypeMeta
+	// +optional
+	metav1.ListMeta
+
+	// Rules specify the audit Level a request should be recorded at.
+	// A request may match multiple rules, in which case the FIRST matching rule is used.
+	// The default audit level is None, but can be overridden by a catch-all rule at the end of the list.
+	Rules []PolicyRule
+}
+
+// PolicyRule maps requests based off metadata to an audit Level.
+// Requests must match the rules of every field (an intersection of rules).
+type PolicyRule struct {
+	// The Level that requests matching this rule are recorded at.
+	Level Level
+
+	// The users (by authenticated user name) this rule applies to.
+	// An empty list implies every user.
+	// +optional
+	Users []string
+	// The user groups this rule applies to. If a user is considered matching
+	// if the are a member of any of these groups
+	// An empty list implies every user group.
+	// +optional
+	UserGroups []string
+
+	// The verbs that match this rule.
+	// An empty list implies every verb.
+	// +optional
+	Verbs []string
+
+	// Rules can apply to API resources (such as "pods" or "secrets"),
+	// non-resource URL paths (such as "/api"), or neither, but not both.
+	// If neither is specified, the rule is treated as a default for all URLs.
+
+	// Resource kinds that this rule matches. An empty list implies all kinds in all API groups.
+	// +optional
+	ResourceKinds []GroupKinds
+	// Namespaces that this rule matches.
+	// The empty string "" matches non-namespaced resources.
```

Comment:
Question to reviewers: does this make sense (`The empty string "" matches non-namespaced resources.`)? Or should we say that the `Namespaces` field is ignored for non-namespaced resources? The argument for the latter is that if I include non-namespaced resources in the `ResourceKinds` I probably meant to catch them, even if I forgot to add the `""` namespace. OTOH, I might want a way to say "log all resources in the namespace foo" without needing to enumerate every namespaced resource.

Actually, now that I wrote it out, I think the second point is a strong argument for keeping it as is (`""` for non-namespaced).